modules/platform/nuxeo-drive-server/nuxeo-drive-mongodb/src/main/java/org/nuxeo/drive/mongodb/MongoDBAuditChangeFinder.java

```diff
@@ -0,0 +1,181 @@
+import org.nuxeo.ecm.core.query.sql.model.OrderByExprs;
+import org.nuxeo.ecm.core.query.sql.model.Predicates;
+import org.nuxeo.ecm.core.query.sql.model.QueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.AuditQueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.AuditReader;
+import org.nuxeo.ecm.platform.audit.api.LogEntry;
+import org.nuxeo.mongodb.audit.MongoDBAuditBackend;
+import org.nuxeo.mongodb.audit.MongoDBAuditEntryReader;
+import org.nuxeo.runtime.api.Framework;
+
+import com.mongodb.client.MongoCollection;
+
+/**
+ * Override the JPA audit based change finder to execute query in BSON.
+ * <p>
+ * The structure of the query executed by the {@link AuditChangeFinder} is:
+ *
+ * <pre>
+ * from LogEntry log where log.repositoryId = :repositoryId
+ *
+ * + AND if ActiveRoots (activeRoots) NOT empty
+ *
+ * from LogEntry log where log.repositoryId = :repositoryId and (
+ * LIST_DOC_EVENTS_IDS_QUERY and ( ROOT_PATHS or COLECTIONS_PATHS) or
+ * (log.category = 'NuxeoDrive' and log.eventId != 'rootUnregistered') )
+ *
+ *
+ * if ActiveRoots EMPTY:
+ *
+ * from LogEntry log where log.repositoryId = :repositoryId and ((log.category =
+ * 'NuxeoDrive' and log.eventId != 'rootUnregistered'))
+ *
+ * + AND (log.id > :lowerBound and log.id <= :upperBound) + order by
+ * log.repositoryId asc, log.eventDate desc
+ * </pre>
+ *
+ * @since 11.3
+ */
+public class MongoDBAuditChangeFinder extends AuditChangeFinder {
+
+    private static final Logger log = LogManager.getLogger(MongoDBAuditChangeFinder.class);
+
+    @Override
+    public long getUpperBound() {
+        RepositoryManager repositoryManager = Framework.getService(RepositoryManager.class);
+        return getUpperBound(new HashSet<>(repositoryManager.getRepositoryNames()));
+    }
+
+    /**
+     * Returns the last available log id in the audit collection considering events older than the last clustering
+     * invalidation date if clustering is enabled for at least one of the given repositories. This is to make sure the
+     * {@code DocumentModel} further fetched from the session using the audit entry doc id is fresh.
+     */
+    @Override
+    public long getUpperBound(Set<String> repositoryNames) {
+        long clusteringDelay = getClusteringDelay(repositoryNames);
+        MongoDBAuditBackend auditService = (MongoDBAuditBackend) Framework.getService(AuditReader.class);
+        QueryBuilder queryBuilder = new AuditQueryBuilder().predicate(
+                Predicates.gt("eventDate", new Date(0)));
+        if (clusteringDelay > -1) {
+            // Double the delay in case of overlapping, see https://jira.nuxeo.com/browse/NXP-14826
+            ZonedDateTime lastClusteringInvalidationDate = ZonedDateTime.now()
+                                                                        .minus(2 * clusteringDelay, ChronoUnit.MILLIS);
+            log.debug("preparing upperBound millis: {}", lastClusteringInvalidationDate);
```

Comment:
I'd say either remove or be more precise, think about the admin who will check the logs.
The message could simply be: "Last clustering invalidation date in milliseconds for the upper bound check: {}"